### PR TITLE
llama : remove LLAMA_MAX_DEVICES and LLAMA_SUPPORTS_GPU_OFFLOAD

### PR DESCRIPTION
**Pull request by ggerganov**
Instead of the defines, use the functions:

- `llama_max_devices()`
- `llama_supports_gpu_offload()`

### CODE REVIEW

**Comment by slaren** [comment]
`server.cpp` also needs to be updated.

**Comment by slaren** [comment]
I assume this is to be able to tell this value at run time, which is useful for example with dynamic linking. In this case, shouldn't `LLAMA_SUPPORTS_GPU_OFFLOAD` also be a function?

**Comment by ggerganov** [comment]
> I assume this is to be able to tell this value at run time, which is useful for example with dynamic linking.

Yes, my main motivation is to reduce GPU-related conditionals in the header files. Nothing specific in mind, just has to be better like this

**Comment by slaren** [comment]
There is still one `LLAMA_SUPPORTS_GPU_OFFLOAD` in `common/train.cpp`

**Comment by slaren** [comment]
If there any downstream applications that depend on these defines to enable GPU offloading, they may break after this change, so it may be a good idea to put a notice.

File: llama.h [hunk 1]
```diff
@@ -5,13 +5,10 @@
...
 #ifdef GGML_USE_CUBLAS
 #include "ggml-cuda.h"
-#define LLAMA_MAX_DEVICES GGML_CUDA_MAX_DEVICES
 #elif defined(GGML_USE_SYCL)
 #include "ggml-sycl.h"
-#define LLAMA_MAX_DEVICES GGML_SYCL_MAX_DEVICES
-#else
-#define LLAMA_MAX_DEVICES 1
-#endif // GGML_USE_CUBLAS
+#endif
```
Comment (slaren) [inline llama.h:10]: I think these includes can be removed now.

**Review by slaren — APPROVED** [review]

**Comment by irthomasthomas** [comment]
Hi, in latest lamma-cpp-python releas (0.2.39) I get this error: `ValueError: Attempt to split tensors that exceed maximum supported devices. Current LLAMA_MAX_DEVICES=1`. Is that related to this change?

**Comment by hamza233** [comment]
> Hi, in latest lamma-cpp-python releas (0.2.39) I get this error: `ValueError: Attempt to split tensors that exceed maximum supported devices. Current LLAMA_MAX_DEVICES=1`. Is that related to this change?

Getting same error

**Comment by slaren** [comment]
It is probably related to this change. `LLAMA_MAX_DEVICES` has been replaced with `llama_max_devices()`.

**Comment by cebtenzzre** [comment]
> It is probably related to this change. `LLAMA_MAX_DEVICES` has been replaced with `llama_max_devices()`.

llama-cpp-python is [calling](https://github.com/abetlen/llama-cpp-python/blob/901827013b732d74f1f67033062d13a6204a62bd/llama_cpp/llama_cpp.py#L103) llama_max_devices() though - after all, it is written in python and gets this value at runtime. I tested it for myself with a CUBLAS build, and get llama_cpp.LLAMA_MAX_DEVICES = 16.

**Comment by slaren** [comment]
It's a weird that there error says `LLAMA_MAX_DEVICES=1` instead of 0 or undefined. Maybe they were using a CPU or Metal build, and this is not really a bug.

**Comment by slaren** [comment]
The current version of llama.cpp will always print the `offloaded x/x layers to GPU` message when using `-ngl`, even in CPU only builds. To be sure that the GPU is actually being used, you should look at the buffer sizes, it should say `CUDA0` instead of `CPU`.

**Comment by hamza233** [comment]
Reinstalling using [this](https://github.com/jllllll/llama-cpp-python-cuBLAS-wheels?tab=readme-ov-file#installation-instructions) worked for me.

**Comment by irthomasthomas** [comment]
Only way I can get it to launch is to roll back llama-cpp-python to v0.2.37

**Comment by thiner** [comment]
@slaren @cebtenzzre I have the same issue with latest llama-cpp-python, is there any workaround I can bypass the LLAMA_MAX_DEVICES=0 issue?

**Pull request merged**